storage/innobase/os/os0file.cc

```diff
@@ -4180,7 +4109,7 @@ bool fil_node_t::read_page0()
 {
 	ut_ad(mutex_own(&fil_system.mutex));
 	const unsigned psize = space->physical_size();
-#ifndef _WIN32
+#ifdef __linux__
```

Comment:
This "linux-only" optimization does not make sense, in terms of performance (I'd like to see benchmark, where it helped), and uses unreliable fstat call (subject of this patch). Can this be removed?

Reply:
The bug that we are trying to address is microsoft/WSL#8443, to my understanding in a Microsoft specific DrvFs driver. You successfully tested a version of the patch where we would invoke the `fstat()` not only on Linux, but on any non-Windows platform. With the above change, we’d invoke `os_file_get_size()` instead of `fstat()` everywhere else but Linux.

With your suggestion, there would be an extra `lseek` system call before the `fstat`, which we will need anyway. This function `fil_node_t::read_page0()` should only be invoked when InnoDB is accessing a tablespace for the very first time, and it needs to determine things like the page size and the file size and whether the file is located on SSD (to disable `innodb_flush_neighbors`, which made sense on HDD storage).

The DrvFs bug scenario that we are trying to prevent is that a file handle was opened with a name, then the file was renamed, and after that, `fstat(2)` would be invoked on the previously invoked file handle. The function `fil_node_t::read_page0()` would typically invoked right after the file was opened for the first time. The caller would set `fil_space_t::size` and therefore prevent any further calls to `read_page0()` for that file. The `fil_space_t` and `fil_node_t` objects would remain in memory until the file is deleted or the database server is shut down. In InnoDB, any `RENAME`, `ALTER`, `OPTIMIZE` or `TRUNCATE` operation that would involve renaming files must be preceded by a call to `fil_node_t::read_page0()`, so that we know that the table is accessible to begin with. There should not be any issue if `fstat` was invoked *before* `rename`. The issue was when `fstat` was invoked *after* `rename`.

Reply:
We do not know much about the nature of this bug https://github.com/microsoft/WSL/issues/8443,

My own testing is limited (I remain a reviewer, not tester in this case), mtr does not run on WSL on Windows partitions, claiming something about "world-writable file my.cnf", so I had to manually try single scenario, and it happened to work, but I can't tell whether it is working OK generally. 

What we know is that MySQL does not use fstat, and that it is working ok. We know that fstat failed in the scenario this patch is aimed to fix.  I do not claim to predict where drvfs will fail. 

We also know, and I measured, that lseek is probably the fastest syscall, and its latency does not count (10 million lseeks in 0.4 sec I measured,  makes 400ns per file which, called once per instance lifetime per file).

lseek is not known to fail. fstat is known to fail. fstat in find_metadata can be completely avoided, if there people are not interested in page_compression (they usually are not), or set innodb_flush_neighbors=0 themselves, rather than relying on helpful innodb introspection. With that , latency of lseek is only 25% of fstat, and if file opening is a performance problem, that Linux-specific handling was aimed to fix, it is best fixed by not doing fstat at all.

Reply:
I see that `fil_node_t::find_metadata()` was already tolerating failures of the `fstat` system call. That is a valid argument for unifying the interface and invoking an extra `lseek` system call. Both `read_page0()` nor `find_metadata()` should be invoked very rarely, when opening or creating a file for the first time.